### ISSUE TITLE
`all` vs `__global__`

### ISSUE THREAD

**Issue by paneq** [RailsEventStore/rails_event_store#131]
- [x] Change `__global__` to `all` in RES_active_record
- [x] Make publishing to `all` work properly (1 record instead of 2 records)
- [x] Keep `all` as higher level detail
- [x] Change the semantic that reading from `all` stream directly will read data written into other streams.
- [x] Change migration to use `all` #129

TLDR: Expose magic `all` stream.

**Comment by mpraglowski** [comment]
Why at all we need to have `__global__` or `all` stream? Isn't the "global" stream a list of all events stored in `event_store_events` with `position`  attribute and without any entry in `event_store_events_in_streams`. Any read/write to `RailsEventStore::GLOBAL_STREAM` should not care about `event_store_events_in_streams`.

> Change the semantic that reading from all stream directly will read data written into other streams.

I think I do not understand this :(

**Comment by paneq** [comment]
@mpraglowski 

> Why at all we need to have `__global__` or `all` stream?

Maybe we don't. That's who I approached it.

>  Isn't the "global" stream a list of all events stored in event_store_events with position attribute and without any entry in event_store_events_in_streams

Perhaps it could be. That would require wrapping saving event and event in streams in a single transaction but maybe that's an interesting design.

> Any read/write to RailsEventStore::GLOBAL_STREAM should not care about event_store_events_in_streams.

I see your vision. My vision was to keep `__global__` implemented in an identical way as all other streams. Thanks for sharing your vision.

**Comment by mpraglowski** [comment]
> I see your vision. My vision was to keep __global__ implemented in an identical way as all other streams. Thanks for sharing your vision.

I like that part about having identical way as all other streams. This is my biggest concern with the way I described it here, it will need some `if` in each method - probably not worth it.

**Comment by pawelpacana** [comment]
Just a thought:

1. let's assume we don't have `__global__` anymore and there is `all` stream with changed semantic (that all events belong to it regardless if an event was published in another named stream or not)
2. we already have an API for reading `read_all_streams_forward` and `read_all_streams_backward`; there is theoretically no need to expose the fact that we have a stream called `all` (i.e. to use it as an argument in `read_events_forward` and `read_events_backward`)
3. all events in `all` stream have no particular position, we determine the order by autoincremented `id` of `event_store_events_in_streams` table

What if:

1. we remove any reference to `GLOBAL_STREAM` in code, resulting in no exposure of `all` stream anymore
2. `event_store_events` has an autoincrement `id` and current `id: :uuid` becomes `event_uuid: :uuid`
3. publishing without stream name appends an event only to `event_store_events`

That would:

1. still allow retrieving all events via `read_all_streams_forward` and `read_all_streams_backward` with a subtle change of depending on autoincrement id of different table (`event_store_events`); events would still be unique

2. simplify some code a bit 

3. might preserve behaviour of explicitly publishing to stream named `all` -- with previous semantics

**Comment by paneq** [comment]
> 2. event_store_events has an autoincrement id and current id: :uuid becomes event_uuid: :uuid

I tried to consciously avoid it. It was trolling for me that events had ID and UUID. And the ID was not ID in fact but a position.

> 3. publishing without stream name appends an event only to event_store_events

We could go in that direction. I preferred a solution where we don't have that split and we explicitly write in `__global__`/`all`.

Also I imagine a different API for reading:

```ruby
read_events.stream("asd").after(uuid).limit(10).order(:forward).to_a
```

That without `stream()` would just read `all/__global__`. Might be more problematic to go in that direction with 2 different tables used for streams, but probably still possible.

If you decide to proceed with your idea know that:

* schema is in 3 different places through the project.
* writing to `event_store_events` is right now outside a transaction (not a problem if event saved but not in any stream). If you follow with this idea, make it inside the transaction.

**Comment by paneq** [comment]
> might preserve behavior of explicitly publishing to stream named all -- with previous semantics

That would be the good part. But I don't see much value in it.

I believe if there are users out there they were not writing explicitly to the `all` stream but rather used the default so they don't care about the stream name. They just wanted to be able to read it later with all events or not even read at all (just trigger the handlers).

**Comment by pawelpacana** [comment]
> My vision was to keep global implemented in an identical way as all other streams. 

> We could go in that direction. I preferred a solution where we don't have that split and we explicitly write in __global__/all.

I supported that direction as well. I however see more and more little things that make "all"/"global" unlike other streams, i.e. https://github.com/RailsEventStore/rails_event_store/pull/138#issuecomment-335746649

At this moment I'd opt for hiding the exposure of global stream. That is assuming you can read all events without specifying stream name.

> I tried to consciously avoid it. It was trolling for me that events had ID and UUID. And the ID was not ID in fact but a position.

I'm fine calling this `global_position`, like eventide does.

**Comment by pawelpacana** [comment]
I'll summarize my ideas with relevant code in separate PR. Thanks for feedback so far!

**Comment by pawelpacana** [comment]
This issue can be now closed since #138 has been merged and #129 updated.
